### ISSUE TITLE
Loading Images in Themes is Broken

### ISSUE THREAD

**Issue by technatica** [leo-editor/leo-editor#808]
The tree boxes in ../leo/Icons/dark and ../leo/themes/dark are transparent and will work with both light and dark themes. But they will not load in Breeze Dark Theme unless the "dark" image folder resides in my home directory (~/dark). 

**EKR**: See the last comment in this issue for a summary of my present thinking.

This is the Leo recommended css to add an image to the background. 

QTextEdit#richTextEdit {
  background-image: url(dark/logo.png);
}

It doesn't work with the "dark" image folder in ~/dark or ~/leo-editor/themes/dark (or ~leo-editor/leo/Icons/dark or ~leo-editor/leo/themes/dark/Icons or several other places like that)

The following does work with ~leo-editor/leo/themes/dark

QTextEdit#richTextEdit {
  background-image: url(/home/chris/leo-editor/leo/themes/dark/logo.png);
}

Interestingly,  if I try to do this

QTextEdit#richTextEdit {
  background-image: url(~/leo-editor/leo/themes/dark/logo.png);
} 

it doesn't work.

Having the boxes pull from ~/dark makes my Leo a happier place, but it does nothing for people who lack such. They get the default Qt boxes, in all their #ffffff glory. 

The desired outcome, I think, would be to support the url(/dark/logo.png) format as a first preference. The path ahead of that should be ../eo-editor/leo/themes/<theme_name_here> period. I cannot think of a good reason to have two locations for images at the same time. A setting would be nice as well so I could change the prepend path to whatever I want for development. ie. ~/dark for development, default (../leo-editor/leo/themes/dark) for deployment. That might be useful, especially if I have several different theming projects on the go.

Sorry it took so long to figure out the boundaries of the problem. It wasn't until I started working on a light theme that I realized the tree boxes were transparent. That made it much easier to track down.

An approach that we know works is the code around color_theme.

Leo Breeze Dark-->Outline indicators-->@string tree-image-closed = dark/branch_closed.svg

populates @tree-image-closed

which can be used in css like this

QTreeView::branch:closed:has-children:!has-siblings,
QTreeView::branch:closed:has-children:has-siblings
{
    border-image: @tree-image-closed;
}

to generate this in the stylesheet

QTreeView::branch:open:has-children:!has-siblings,
QTreeView::branch:open:has-children:has-siblings
{
    border-image: url(/home/chris/leo-editor/themes/dark/branch_closed.svg);
}

to pull the correct image from the correct directory (correct IMHO) and display it.

**Comment by edreamleo** [comment]
On Tue, Mar 13, 2018 at 2:31 PM, Chris George <notifications@github.com>
wrote:

> The tree boxes in ../leo/Icons/dark and ../leo/themes/dark are transparent
> and will work with both light and dark themes. But they will not load in
> Breeze Dark Theme unless the "dark" image folder resides in my home
> directory (~/dark).
>

​Thanks for this. There are path-related computations all over Leo, and all
over the theme code.  Making them consistent is like herding cats.

Edward

**Comment by technatica** [comment]
I would replace the schemes in both 

`qt_gui.getImageImageFinder`

and

`ssm.expand_css_constants & helpers`

with a scheme that makes CWD equivalent to loadDir, assuming there are no unintended consequences.

We should also support the Qt syntax for an image url.

`url(:/<theme_name>/<image_name>);`

Tree icon images belong to the style and should be treated the same as all other style images. If they do not exist, the failsafe is the built-in Qt boxes that we see now.

In my opinion, a theme should entirely handle the UI. Images and stylesheets should have exactly one source. If that source does not exist, there needs to be a sane default theme/style that can be loaded.

The logical places for theme images is in ../leo/themes/<theme_name>. No Icons folder, just the root. 
 
Alternatively, we could achieve the same goal by using the existing @constants mechanism used by the open and close arrows. Generalize it so that all `@string <constant> = <theme_name>/<image_name>` in the "Outline indicators" node assigns the image to the constant and @<constant> is used in the stylesheet nodes for all images.

I still do not know where the actual assignment of box00.png (box01.png, box02.png etc.) to substitute for the default Qt box is made in Leo's code. Will this support bmp, jpg, svg etc. for tree boxes? I don't know where to look. Searching for "box0" in LeoPyRef.leo doesn't go far.

As far as changes to the existing code, I wouldn't really know where to begin.

**Comment by edreamleo** [comment]
On Thu, Mar 15, 2018 at 1:33 PM, Chris George <notifications@github.com>
wrote:

> I would replace the schemes in both
>
> qt_gui.getImageImageFinder
>
> and
>
> ssm.expand_css_constants & helpers
>
> with a scheme that makes CWD equivalent to loadDir, assuming there are no
> unintended consequences.
>
​Leo typically bases paths on g.app.loadDir, and the two methods you
mention are no exceptions.  ​Opening dialogs can change the cwd, but that's
a separate issue, imo.

> We should also support the Qt syntax for an image url.
>
> url(:/<theme_name>/<image_name>);
>
Tree icon images belong to the style and should be treated the same as all
> other style images. If they do not exist, the failsafe is the built-in Qt
> boxes that we see now.
>
> In my opinion, a theme should entirely handle the UI. Images and
> stylesheets should have exactly one source. If that source does not exist,
> there needs to be a sane default theme/style that can be loaded.
>
> The logical places for theme images is in ../leo/themes/<theme_name>. No
> Icons folder, just the root.
>
> Alternatively, we could achieve the same goal by using the existing
> @Constants <https://github.com/constants> mechanism used by the open and
> close arrows. Generalize it so that all @string <constant> =
> <theme_name>/<image_name> in the "Outline indicators" node assigns the
> image to the constant and @ is used in the stylesheet nodes for all images.
>
​I am going to put these comments on the back burner for awhile.  I'm
pretty sure I understand the issues involved. I'll try to generalize the
code as much as possible.

Edward

**Comment by technatica** [comment]
Testing after 7b6057cb46e6.

The following urls both work to bring in an image as a background. 

`url(../leo-editor/leo/Icons/dark/logo.png)`
`url(../leo-editor/leo/themes/dark.logo.png)`

If I put a "dark" directory in my home directory (~/dark), `url(..dark/logo.png)` works fine. If I remove the dark directory, no image is found.

The transparent box icons for the tree remain transparent. They are being sourced from ../leo/theme/dark, for sure. I renamed the directory and white icons load, but only after restarting Leo. Several save theme, "reload-style-sheets, open theme did not change the icons, but a restart does.

All good so far, some clues I hope.

Then I used "File, Open" to open a file in an unrelated folder.

The boxes are good. They still load. But the `url(../leo-editor/leo/themes/dark.logo.png)` url no longer brings in the image even after several save theme, "reload-style-sheets, open theme cycles. Restarting Leo puts the "File, Open" dialog back to ~/leo-editor and the url works again, until I open a file from somewhere else. 

Then the image in all current files vanishes until I click into the body pane. Then it appears. You have to see this to believe it.

So, open Leo. Have a file open so Leo will remain open. Open theme file, add above url pointing to an image that exists in a folder somewhere in the ../leo-editor directory. Save theme file, reload-style-sheets, open theme file. Background loads. File, open a Leo file that exists somehwere not in the leo-editor directory. For me the image disappears. Then I click into the body of any open file and it appears.

I will report observations as I investigate further. I think there is more than one thing going on here and their interaction is muddling things.

**Comment by edreamleo** [comment]
I am going to close this issue.  Recent work generalizes the search for icons, etc.

Please reopen if problems remain.

**Comment by technatica** [comment]
Summary:

I can access an image using url(../dark/logo.png) if the directory ~/dark exists. Remove that directory and no images load.

The following syntaxes do not work: 
url(:/dark/logo.png) *standard Qt syntax*
url(dark/logo.png) *truncated path as used by color_theme to source the open/closed images*

The only generalization that works is url(../leo-editor/leo/themes/dark/logo.png)

This works with any path as long as it is below leo-editor in the filesystem. If this structure applies to all three platforms (Win, Mac,Linux) we're all good and I will proceed to use this structure to style with images.

But opening any file from anywhere in the filesystem (including leo-editor) breaks the background-image in the currently opened files and all other subsequently opened files, even though the background-image is specified in myLeoSettings.leo. Would you like this opened as a different issue?

**Comment by edreamleo** [comment]
On Mon, Mar 19, 2018 at 11:18 AM, Chris George <notifications@github.com>
wrote:

> Summary:
>
​[snip]

> But opening any file from anywhere in the filesystem (including
> leo-editor) breaks the background-image in the currently opened files and
> all other subsequently opened files, even though the background-image is
> specified in myLeoSettings.leo.
>
​Good to know.​


> Would you like this opened as a different issue?
>
​No.  The present issue is fine.

Edward

**Comment by edreamleo** [comment]
Pretty much everything related to qt_gui.getImageFinder must be revised to deal with themes.

In particular, LeoQtGui can't possibly use a local color_theme ivar.  Instead, it must (somehow) access color_theme setting in the theme file or myLeoSettings.leo.

Clearly this is possible, but it may take some doing.

**Comment by edreamleo** [comment]
To summarize this issue:

Relative url's in stylesheets are problematic. According to [this page](http://doc.qt.io/qt-5/resources.html#resource-collection-files-op-op-qrc), the url(:/...)` syntax refers to resource collection (.qrc) files. Such files do not exist in Leo.  What happens in this case seems to be undefined/unexplained.  It may involve the cwd.

Imo, the *only* reasonable solution is for ssm.replace_indicator_constants (or a new helper) to replace:

    url(:/whatever)

by:

    url(<full path to>whatever)

This will likely involve *searching* for "whatever" in a list of "usual places".  Such lists already exist, but may need refining.

**Note**: ssm.replace_indicator_constants calls ssm.replace_indicator_constants to handle the following two specific patterns:

    image: @tree-image-closed
    image: @tree-image-open

The code in ssm.replace_indicator_constants is very specific, which may be why it works ;-) Generalizing ssm.replace_indicator_constants could be said to be the heart of this issue.

**Other methods are not involved**

Leo's code calls qt_gui.getImageFinder *only* to resolve icon boxes. Nothing else.

qt_gui.getImageFinder should be revised to deal with themes. However, that does not seem urgent, and does not seem to be related to this issue.

**Comment by edreamleo** [comment]
Rev 168d79c may fix this.  Please reopen if problems remain.

**Comment by edreamleo** [comment]
I am going to reopen this issue.  Some recent rev broke Terry's theme.  That's unacceptable. We can't have the new code break existing themes.

**Comment by edreamleo** [comment]
Rev fc3f40b8f completes the work for now.

qt_gui.getImageFinder works *provided* that theme.leo files should set the @string theme_name to the name of the theme file itself.
